tests/experimental/pydantic_first_class/test_basic.py

```diff
@@ -0,0 +1,536 @@
+    assert definition.name == "User"
+
+    [field1, field2] = definition.fields
+
+    assert field1.python_name == "age_"
+    assert field1.type is int
+    assert field1.graphql_name == "age"
+
+    assert field2.python_name == "password"
+    assert isinstance(field2.type, StrawberryOptional)
+    assert field2.type.of_type is str
+
+
+def test_union():
+    @first_class()
+    class BranchAType(pydantic.BaseModel):
+        field_a: str
+
+    @first_class()
+    class BranchBType(pydantic.BaseModel):
+        field_b: int
+
+    @first_class()
+    class UserType(pydantic.BaseModel):
+        age: int
+        union_field: Union[BranchAType, BranchBType]
+
+    definition: StrawberryObjectDefinition = UserType.__strawberry_definition__
+    assert definition.name == "UserType"
+
+    [field1, field2] = definition.fields
+
+    assert field1.python_name == "age"
+    assert field1.type is int
+
+    assert field2.python_name == "union_field"
+    assert isinstance(field2.type, StrawberryUnion)
+    assert field2.type.types[0] is BranchAType
+    assert field2.type.types[1] is BranchBType
+
+
+def test_enum():
+    @strawberry.enum
+    class UserKind(Enum):
+        user = 0
+        admin = 1
+
+    @first_class()
+    class UserType(pydantic.BaseModel):
+        age: int
+        kind: UserKind
+
+    definition: StrawberryObjectDefinition = UserType.__strawberry_definition__
+    assert definition.name == "UserType"
+
+    [field1, field2] = definition.fields
+
+    assert field1.python_name == "age"
+    assert field1.type is int
+
+    assert field2.python_name == "kind"
+    assert isinstance(field2.type, EnumDefinition)
+    assert field2.type.wrapped_cls is UserKind
+
+
+def test_interface():
+    @first_class(is_interface=True)
+    class BaseType(pydantic.BaseModel):
+        base_field: str
+
+    @first_class()
+    class BranchAType(BaseType):
+        field_a: str
+
+    @first_class()
+    class BranchBType(BaseType):
+        field_b: int
+
+    @first_class()
+    class UserType(pydantic.BaseModel):
+        age: int
+        interface_field: BaseType
+
+    definition: StrawberryObjectDefinition = UserType.__strawberry_definition__
+    assert definition.name == "UserType"
+
+    [field1, field2] = definition.fields
+
+    assert field1.python_name == "age"
+    assert field1.type is int
+
+    assert field2.python_name == "interface_field"
+    assert field2.type is BaseType
+
+
+# Excepted failure. Need to fix this. Currently fails because we now mutate the same
+# __strawberry_definition__ field on the pydantic model. This is a problem because
+# we need to be able to have both a) an input type and b) an output type on the
+# same pydantic model. Maybe we need to have a separate __strawberry_input_definition__
+# Or just store them in a dict rather than on the model itself.
+@pytest.mark.xfail(reason="Need to fix this")
+def test_both_output_and_input_type():
+    @first_class(name="WorkInput", is_input=True)
```

Comment:
One problem is when your pydantic type needs to be both a Input and Output type. Now you can't mark them with the same decorator - because that will override the existing `__strawberry_definition__`.